### PR TITLE
Balder/create diversity on demand

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe PR

### CODE REVIEW

**Comment by baldersheim** [comment]
@havardpe Now it should be complete.

File: searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp [hunk 1]
```diff
@@ -37,6 +41,40 @@ void tag_match_data(const HandleRecorder::HandleSet &handles, MatchData &match_d
...
+                                                   limit_max_filter_coverage, samplePercentage, postFilterMultiplier);
+    } else if (rankSetup.hasMatchPhaseDegradation()) {
+        return std::make_unique<DegradationParams>(rankSetup.getDegradationAttribute(), rankSetup.getDegradationMaxHits(),
+                                                   !rankSetup.isDegradationOrderAscending(), rankSetup.getDegradationMaxFilterCoverage(),
+                                                   rankSetup.getDegradationSamplePercentage(), rankSetup.getDegradationPostFilterMultiplier());
+    }
+    return std::unique_ptr<DegradationParams>();
+}
+
+DiversityParams
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp]: it feels wrong to not use the same strategy for both paramer types...

File: searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp [hunk 2]
```diff
@@ -191,4 +212,19 @@ MatchToolsFactory::createMatchTools() const
                                         *_match_limiter, _queryEnv, _mdl, _rankSetup, _featureOverrides);
 }
 
+std::unique_ptr<IDiversifier> MatchToolsFactory::createDiversifier() const
+{
+    if (_diversityParams._attribute.empty()) {
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp]: should also check min_groups here

**Review by havardpe — COMMENTED** [review]
looks good overall, please consider my comments

File: searchcore/src/tests/proton/matching/matching_test.cpp [hunk 3]
```diff
@@ -329,6 +340,12 @@ struct MyWorld {
         { return SearchReply::UP(); }
     };
 
+    MatchToolsFactory::UP get_mtf(SearchRequest::SP req) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/tests/proton/matching/matching_test.cpp]: Fixed

File: searchcore/src/tests/proton/matching/matching_test.cpp [hunk 4]
```diff
@@ -528,6 +545,78 @@ TEST("require that re-ranking is performed (multi-threaded)") {
...
+using namespace search::fef::indexproperties::matchphase;
+TEST("require that re-ranking is diverse") {
+    MyWorld world;
+    world.basicSetup();
+    world.setupSecondPhaseRanking();
+    world.basicResults();
+    SearchRequest::SP request = world.createSimpleRequest("f1", "spread");
+    auto mtf = world.get_mtf(request);
+    auto diversity = mtf->createDiversifier();
+    EXPECT_FALSE(diversity);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/tests/proton/matching/matching_test.cpp]: Fixed

File: searchcore/src/tests/proton/matching/matching_test.cpp [hunk 5]
```diff
@@ -528,6 +545,78 @@ TEST("require that re-ranking is performed (multi-threaded)") {
...
+    auto & rankProperies = request->propertiesMap.lookupCreate(MapNames::RANK);
+    rankProperies.add(DiversityAttribute::NAME, "a2")
+                 .add(DiversityMinGroups::NAME, "3")
+                 .add(DiversityCutoffStrategy::NAME, "strict");
+    mtf = world.get_mtf(request);
+    diversity = mtf->createDiversifier();
+    EXPECT_TRUE(diversity);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/tests/proton/matching/matching_test.cpp]: Fixed

File: searchcore/src/tests/proton/matching/matching_test.cpp [hunk 6]
```diff
@@ -528,6 +545,78 @@ TEST("require that re-ranking is performed (multi-threaded)") {
...
+    EXPECT_EQUAL(1400.0, reply->hits[2].metric);
+    EXPECT_EQUAL(document::DocumentId("doc::600").getGlobalId(),  reply->hits[3].gid);
+    EXPECT_EQUAL(600.0, reply->hits[3].metric);
+    EXPECT_EQUAL(document::DocumentId("doc::500").getGlobalId(),  reply->hits[4].gid);
+    EXPECT_EQUAL(500.0, reply->hits[4].metric);
+    EXPECT_GREATER(world.matchingStats.matchTimeAvg(), 0.0000001);
+    EXPECT_GREATER(world.matchingStats.rerankTimeAvg(), 0.0000001);
+}
+
+TEST("require that re-ranking is forced diverse") {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/tests/proton/matching/matching_test.cpp]: Fixed

File: searchcore/src/vespa/searchcore/proton/matching/match_phase_limiter.h [hunk 7]
```diff
@@ -69,6 +68,41 @@ struct NoMatchPhaseLimiter : MaybeMatchPhaseLimiter {
...
+    DiversityParams() : DiversityParams("", 0, 0, CutoffStrategy::LOOSE) { }
+    DiversityParams(const vespalib::string & attribute, uint32_t min_groups,
+                    double cutoff_factor, CutoffStrategy cutoff_strategy)
+        : _attribute(attribute),
+          _min_groups(min_groups),
+          _cutoff_factor(cutoff_factor),
+          _cutoff_strategy(cutoff_strategy)
+    { }
+
+    vespalib::string  _attribute;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/match_phase_limiter.h]: Fixed

File: searchcore/src/vespa/searchcore/proton/matching/match_phase_limiter.h [hunk 8]
```diff
@@ -69,6 +68,41 @@ struct NoMatchPhaseLimiter : MaybeMatchPhaseLimiter {
...
+    DegradationParams(const vespalib::string &attribute, size_t max_hits, bool descending, double max_filter_coverage,
+                      double sample_percentage, double post_filter_multiplier)
+        : _attribute(attribute),
+          _max_hits(max_hits),
+          _descending(descending),
+          _max_filter_coverage(max_filter_coverage),
+          _sample_percentage(sample_percentage),
+          _post_filter_multiplier(post_filter_multiplier)
+    { }
+    vespalib::string _attribute;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/match_phase_limiter.h]: Fixed

File: searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp [hunk 9]
```diff
@@ -37,6 +41,40 @@ void tag_match_data(const HandleRecorder::HandleSet &handles, MatchData &match_d
     }
 }
 
+std::unique_ptr<DegradationParams>
+getDegradationParams(const RankSetup & rankSetup, const Properties & rankProperties)
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp]: Fixed

File: searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp]: Fixed

File: searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/matching/match_tools.cpp]: Fixed

**Comment by baldersheim** [comment]
@havardpe All comments addressed, hopfully to you liking

File: searchcore/src/tests/proton/matching/matching_test.cpp [hunk 10]
```diff
@@ -528,6 +545,87 @@ TEST("require that re-ranking is performed (multi-threaded)") {
...
+    world.basicResults();
+    SearchRequest::SP request = world.createSimpleRequest("f1", "spread");
+    auto mtf = world.create_mtf(request);
+    auto & rankProperies = request->propertiesMap.lookupCreate(MapNames::RANK);
+    rankProperies.add(DiversityAttribute::NAME, "a2")
+                 .add(DiversityMinGroups::NAME, "3")
+                 .add(DiversityCutoffStrategy::NAME, "strict");
+    mtf = world.create_mtf(request);
+    auto diversity = mtf->createDiversifier();
+    EXPECT_TRUE(diversity);
```
Comment (havardpe) [inline searchcore/src/tests/proton/matching/matching_test.cpp]: I think activation check should also be put in a separate test

File: searchcore/src/tests/proton/matching/matching_test.cpp [hunk 11]
```diff
@@ -528,6 +545,87 @@ TEST("require that re-ranking is performed (multi-threaded)") {
...
+    auto mtf = world.create_mtf(request);
+    auto diversity = mtf->createDiversifier();
+    EXPECT_FALSE(diversity);
+    auto & rankProperies = request->propertiesMap.lookupCreate(MapNames::RANK);
+    rankProperies.add(DiversityAttribute::NAME, "a3")
+                 .add(DiversityMinGroups::NAME, "3")
+                 .add(DiversityCutoffStrategy::NAME, "strict");
+    mtf = world.create_mtf(request);
+    diversity = mtf->createDiversifier();
+    EXPECT_TRUE(diversity);
```
Comment (havardpe) [inline searchcore/src/tests/proton/matching/matching_test.cpp]: activation check in a separate test would reduce redundant testing...

File: searchcore/src/tests/proton/matching/matching_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (havardpe) [reply searchcore/src/tests/proton/matching/matching_test.cpp]: This is still duplicated inside the other tests...

**Comment by havardpe** [comment]
overall: 👍

File: searchcore/src/tests/proton/matching/matching_test.cpp (hunk 10 above)
Comment (baldersheim) [reply searchcore/src/tests/proton/matching/matching_test.cpp]: Fixed in #6546

File: searchcore/src/tests/proton/matching/matching_test.cpp (hunk 11 above)
Comment (baldersheim) [reply searchcore/src/tests/proton/matching/matching_test.cpp]: Fixed in #6546

**Pull request merged**